Update image spec to support multiple images, add build tooling for bytecode images (#1141)
bpfman/bpfman: docs/developer-guide/shipping-bytecode.md

To provide relevant metadata regarding the bytecode to any consumers, some relevant labels
**MUST** be defined on the image.

These labels are dynamic and defined as follows:

> - `io.ebpf.programs`: A label which defines The eBPF programs stored in the bytecode image.

Reviewer: Is it possible (as a future enhancement) to automatically pull the lists of programs and maps from the bytecode?

Reviewer: Strictly speaking, I don't think they are required for proper operation as we've demonstrated with the multi-program BpfApplication images.  I think our code just uses them as a sanity check when loading.  If we didn't do this check, I think the Aya code would barf if the user tried to load something that didn't exist in the image.  However, I think they are useful for documenting what is contained in the image.

Author: No enhancement needed `bpfman image build` already does this (added in this PR)!! 🥳

Reviewer: I guess I missed that.

>    the value of the label is a list which must contain a valid JSON object with

Reviewer 2: ```suggestion
   The value of the label is a list which must contain a valid JSON object with
```

   Key's specifying the program name, and values specifying the program type i.e:
   "{ "pass" : "xdp" , "counter" : "tc", ...}".

> - `io.ebpf.maps`: A label which defines The eBPF maps stored in the bytecode image.

Reviewer 2: ```suggestion
- `io.ebpf.maps`: A label which defines the eBPF maps stored in the bytecode image.
```

   the value of the label is a list which must contain a valid JSON object with
   Key's specifying the map name, and values specifying the map type i.e:
   "{ "xdp_stats_map" : "per_cpu_array", ...}".

### Building a Backwards compatible OCI compliant image

Bpfman does not provide wrappers around compilers like clang since many eBPF
libraries (i.e aya, libbpf, cilium-ebpf) already do so, meaning users are expected
to pass in the correct ebpf program bytecode for the appropriate platform. However,
bpfman does provide a few image builder commands to make this whole process easier.

> An Example Containerfile can be found at `/packaging/container/deployment/Containerfile.bytecode`

Reviewer 2: Is this `bpfman/Containerfile.bytecode` now?

Author: It's bpfman/Containerfile.bytecode and bpfman/Containerfile.bytecode.multi.arch.

However in the future it will be no containerfile 🥳

#### Host Platform Architecture Image Build

```console
bpfman image build -b ./examples/go-xdp-counter/bpf_bpfel.o -f Containerfile.bytecode --tag quay.io/<USER>/go-xdp-counter
```

> Where `./examples/go-xdp-counter/bpf_x86_bpfel.o` is the directory the bytecode object file is located.

Reviewer: ```suggestion
Where `./examples/go-xdp-counter/bpf_x86_bpfel.o` is the path to the bytecode object file.
```

  configuration layer (`application/vnd.oci.image.config.v1+json`) of the image.

```bash
skopeo inspect docker://quay.io/astoycos/go-xdp-counter
{

>     "Name": "quay.io/astoycos/go-xdp-counter",

Author: remove astoycos
